### DISCUSSION TITLE
Concurrency with pybricks

### DISCUSSION THREAD

**Discussion by wedgebo** [pybricks/support#1272] [Q&A] [answered]
How would I realize concurrency with pybricks. 
I'd like to run a spike prime hub with pybricks but did not find uasyncio module in the documentation.
pybricks looks so promising but I cannot imagine to not using concurrency.

**Comment by laurensvalk** [comment, marked as answer]
Try https://beta.pybricks.com/

In the docs, under the tools module, you’ll find multitasking tools using async/await along with an example.

Does that allow you to do what you want?

**Reply by cegekaJG** [reply]
Did this get removed? I can't see anything in the `tools` module of Pybricks, and there's nothing in the Micropython modules either.
`uselect` mentions using `ipoll` for asynchronous functions, is there an example for that?
Edit: Nvm, I was prompted to load the latest version and the `multitask` function was here this time.

**Comment by wedgebo** [comment]
Thank you.  The multitasking tools in this module look like a subset of uasyncio and are, of course, valuable. I wonder why I did not manage to find this module.
Sadly something similar to uasyncio.Event seems to be missing. Or, at least, I am too blind to find them again. I will definitly keep browsing the docs.

**Reply by laurensvalk** [reply]
It is indeed like a subset. You may be able to implement things like events based on the provided tools.

You can make this in a separate file if you like, and then simply import them in your main script.

You could even make decorators and the like to mimic behaviors like running functions triggered by certain conditions.

**Comment by Nashatyrev** [comment]
Thanks guys! 
Isn't the concurrency available in the original Lego firmware yet?

**Reply by laurensvalk** [reply]
Possibly, but documention may be scarce or invalid completely because the official spike python api completely changed recently and mindstorms has been discontinued altogether.

We try to keep things really constant at Pybricks, by adding new features without breaking the old ones.
